vaas/vaas/metrics/prometheus.py

```diff
@@ -0,0 +1,54 @@
+import logging
+
+from datetime import timedelta
+from typing import Optional, TypedDict, Union
+from django.conf import settings
+from prometheus_client import CollectorRegistry, Gauge, Summary, push_to_gateway
+
+from vaas.metrics.models import Metrics
+
+logger = logging.getLogger(__name__)
+
+class MetricsBucket(TypedDict, total=False):
+    name: str
+    metric: Union[Summary, Gauge]
+
+
+class PrometheusClient:
+    def __init__(self, registry: CollectorRegistry) -> None:
+        self.host: str = f'{settings.PROMETHEUS_GATEWAY_HOST}:{settings.PROMETHEUS_GATEWAY_PORT}'
+        self.job: str = settings.PROMETHEUS_GATEWAY_JOB
+        self.registry: CollectorRegistry = registry
+
+    def push(self) -> None:
+        try:
+            push_to_gateway(gateway=self.host, job=self.job, registry=self.registry)
+        except Exception as e:
+            logger.exception(f'PrometheusClient: cannot push metric to vmagent')
+            pass
+
+
+class PrometheusMetrics(Metrics):
+    def __init__(self) -> None:
+        self.metrics_bucket: MetricsBucket = {}
+        self.registry: CollectorRegistry = CollectorRegistry()
+        super().__init__(
+            client=PrometheusClient(self.registry))
+
+    def _get_or_create(self, name: str, type: Union[Summary, Gauge]) -> Optional[Union[Summary, Gauge]]:
+        name_with_suffix: str = f'{name}_microseconds'
+        metric = self.metrics_bucket.get(name_with_suffix)
+        if not metric:
+            self.metrics_bucket[name_with_suffix] = type(name_with_suffix, name, registry=self.registry) # type: ignore
+            return self.metrics_bucket.get(name_with_suffix)
+        return metric
+
+    def sum(self, metric_name: str, value: timedelta) -> None:
+        if metric_name:
+            self._get_or_create(metric_name, Summary).observe(value.microseconds) # type: ignore
+            self.client.push()
+
+    def gauge(self, metric_name: str, value: Union[int, float]) -> None:
+        if metric_name:
+            self._get_or_create(metric_name, Gauge).set(value) # type: ignore
+            self.client.push()
```

Comment:
Where are the tests for that?

Reply:
Yes, I know there are no tests. I'm making it in another PR.